src/valkey-cli.c

```diff
@@ -218,6 +218,10 @@ static struct config {
     int shutdown;
     int monitor_mode;
     int pubsub_mode;
+    int pubsub_channel_count;
+    int pubsub_pattern_count;
+    int pubsub_shard_count;
```

Comment:
There is something strange with how the subscriptions are counted. It seems like channels and patterns are counted together with one counter while only shard-channels are counted with a separate counter.

I tried this just now:

```
127.0.0.1:6379> subscribe ch
1) "subscribe"
2) "ch"
3) (integer) 1
127.0.0.1:6379(subscribed mode)> psubscribe p*
1) "psubscribe"
2) "p*"
3) (integer) 2
127.0.0.1:6379(subscribed mode)> ssubscribe s
1) "ssubscribe"
2) "s"
3) (integer) 1
127.0.0.1:6379(subscribed mode)> unsubscribe
1) "unsubscribe"
2) "ch"
3) (integer) 1
127.0.0.1:6379(subscribed mode)> punsubscribe
1) "punsubscribe"
2) "p*"
3) (integer) 0
```

If you keep three counters internally, I think you can fail to detect when unsubscribe is finished. If you use all three in the same test case, I think you can see this behaviour.

Reply:
I found this issue as well when I was manually testing the code, currently looking in this issue.

Reply:
Isn't the solution to use two counters instead of three?

    int pubsub_unsharded_count; /* channels and patterns */
    int pubsub_sharded_count;   /* shard channels */

Reply:
Yes I think you're correct, i've pushed those changes. I have another question. Are we allowed to use punsubscribe to unsubscribe from channels we have used the subscribe command on?

for example:
```subscribe ch1 ch2 ch3
punsubscribe ch*
```

Reply:
I believe patterns and channels are stored separately on the server. If you `subscribe ch1` and `psubscribe ch*`, then you will get messages on `ch` until you do both `unsubscribe` and `punsubscribe`. That's what I think, but I haven't tested this.

Reply:
When I tested it that's what I got as well. We had to do unsubscribe and punsubscribe to fully unsubscribe from the channels and get out of subscribe mode. I assume this is the desired functionality.